spock-core/src/main/java/spock/util/concurrent/PollingConditions.java

```diff
@@ -148,34 +148,43 @@ public void eventually(Closure<?> conditions) throws InterruptedException {
    * @throws InterruptedException if evaluation is interrupted
    */
   @ConditionBlock
-  public void within(double seconds, Closure<?> conditions) throws InterruptedException  {
+  public void within(double seconds, Closure<?> conditions) throws InterruptedException {
     long timeoutMillis = toMillis(seconds);
     long start = System.currentTimeMillis();
     long lastAttempt = 0;
-    Thread.sleep(toMillis(initialDelay));
+    // Thread.sleep(0) would yield the processor to higher-priority threads,
+    // which would slow down a polling condition without initial delay unnecessarily
+    if (initialDelay > 0) {
+      Thread.sleep(toMillis(initialDelay));
+    }
 
     long currDelay = toMillis(delay);
     int attempts = 0;
+    long elapsedTime = 0;
+    Throwable testException = null;
 
-    while(true) {
+    while (elapsedTime <= timeoutMillis) {
       try {
-        attempts++;
-        lastAttempt = System.currentTimeMillis();
-        GroovyRuntimeUtil.invokeClosure(conditions);
+        try {
+          GroovyRuntimeUtil.invokeClosure(conditions);
+        } finally {
+          lastAttempt = System.currentTimeMillis();
+          attempts++;
+        }
         return;
       } catch (Throwable e) {
-        long elapsedTime = lastAttempt - start;
-        if (elapsedTime >= timeoutMillis) {
-          String msg = String.format("Condition not satisfied after %1.2f seconds and %d attempts", elapsedTime / 1000d, attempts);
-          throw new SpockTimeoutError(seconds, msg, e);
-        }
+        testException = e;
+        elapsedTime = lastAttempt - start;
         final long timeout = Math.min(currDelay, start + timeoutMillis - System.currentTimeMillis());
```

Comment:
My intention was to improve the behaviour of what I perceived as a problematic or even buggy piece of code, measuring the wrong thing and using `Thread.sleep(0)` unnecessarily, giving up control over the thread for no reason. In order to do that, I applied a minimal change. I did not intend to change an unproblematic line of code which was out of scope for my improvement. The boyscout rule for refactoring is: "Leave the camp ground behind cleaner than you found it." That does not mean I have to refactor the whole class or even the whole method. Actually, it does not matter that the variable is shadowing the field because it poses no problem here.

Having said that - I want you to understand my way of working - of course I can rename the variable for you.